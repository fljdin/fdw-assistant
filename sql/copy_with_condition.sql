-- public.t1 contains negative values that we wont copy
CREATE TABLE public.t1 (
    id serial primary key,
    value smallint not null
);

INSERT INTO public.t1 (id, value)
    SELECT i, i - 100 FROM generate_series(1, 199) i;

CREATE TABLE public.t2 (
    id serial primary key,
    value smallint not null CHECK (value >= 0)
);

INSERT INTO config (source, target, pkey, priority, parts, trunc, condition, batchsize) VALUES
-- t1 will be copied with a condition that filters negative values
    ('public.t1', 'public.t2', 'id', 200, 2, true, 'value >= 0', null);

SELECT * FROM config WHERE target = 'public.t2'::regclass;

SELECT invocation FROM plan('{public.t2}');

-- copy(1) and copy(2) should copy the positive values from source.t1
-- only job with part #0 should truncate the target table
CALL copy(1);
CALL copy(2);

-- public.t2 should have 100 rows
SELECT count(*) FROM public.t2;
