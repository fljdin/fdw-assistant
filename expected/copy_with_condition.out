-- public.t1 contains negative values that we wont copy
CREATE TABLE public.t1 (
    id serial primary key,
    value smallint not null
);
CREATE TABLE
INSERT INTO public.t1 (id, value)
    SELECT i, i - 100 FROM generate_series(1, 199) i;
INSERT 0 199
CREATE TABLE public.t2 (
    id serial primary key,
    value smallint not null CHECK (value >= 0)
);
CREATE TABLE
INSERT INTO config (source, target, pkey, priority, parts, trunc, condition, batchsize) VALUES
-- t1 will be copied with a condition that filters negative values
    ('public.t1', 'public.t2', 'id', 200, 2, true, 'value >= 0', null);
INSERT 0 1
SELECT * FROM config WHERE target = 'public.t2'::regclass;
  source   |  target   | pkey | priority | parts | trunc | condition  | batchsize 
-----------+-----------+------+----------+-------+-------+------------+-----------
 public.t1 | public.t2 | id   |      200 |     2 | t     | value >= 0 |          
(1 row)

SELECT invocation FROM plan('{public.t2}');
       invocation        
-------------------------
 CALL assistant.copy(1);
 CALL assistant.copy(2);
(2 rows)

-- copy(1) and copy(2) should copy the positive values from source.t1
-- only job with part #0 should truncate the target table
CALL copy(1);
NOTICE:  Executing: TRUNCATE public.t2
NOTICE:  Executing: SELECT count(id) FROM public.t1 WHERE id > 0 AND id % 2 = 0 AND value >= 0
NOTICE:  Executing: INSERT INTO public.t2 SELECT id, value FROM public.t1 WHERE id > 0 AND id % 2 = 0 AND value >= 0 ORDER BY id 
CALL
CALL copy(2);
NOTICE:  Executing: SELECT count(id) FROM public.t1 WHERE id > 0 AND id % 2 = 1 AND value >= 0
NOTICE:  Executing: INSERT INTO public.t2 SELECT id, value FROM public.t1 WHERE id > 0 AND id % 2 = 1 AND value >= 0 ORDER BY id 
CALL
-- public.t2 should have 100 rows
SELECT count(*) FROM public.t2;
 count 
-------
   100
(1 row)

